compiler-core/src/type_/expression.rs

```diff
@@ -1062,24 +1062,72 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
                 container,
                 index: _,
                 type_: (),
-            } => {
-                let container = self.infer_clause_guard(*container)?;
-                let container = Box::new(container);
-                let container_type = container.type_();
-                let (index, label, type_) = self.infer_known_record_access(
-                    container_type,
-                    container.location(),
-                    FieldAccessUsage::Other,
-                    location,
-                    label,
-                )?;
-                Ok(ClauseGuard::FieldAccess {
-                    container,
-                    label,
-                    index: Some(index),
-                    location,
-                    type_,
-                })
+            } => match self.infer_clause_guard(*container.clone()) {
```

Comment:
Finding this function quite hard to read with its mix of different concerns and very large amount of nesting. Could you extract some functions from this? Perhaps one for records and one for modules.